gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/AdvanceDagProc.java

```diff
@@ -0,0 +1,134 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration.proc;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.Future;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Maps;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.runtime.api.JobSpec;
+import org.apache.gobblin.runtime.api.Spec;
+import org.apache.gobblin.runtime.api.SpecProducer;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.orchestration.DagManagerUtils;
+import org.apache.gobblin.service.modules.orchestration.DagProcFactory;
+import org.apache.gobblin.service.modules.orchestration.TimingEventUtils;
+import org.apache.gobblin.service.modules.orchestration.task.AdvanceDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+
+import static org.apache.gobblin.service.ExecutionStatus.RUNNING;
+
+
+/**
+ * An implementation of {@link DagProc} dealing with advancing to the next node in the {@link Dag}.
+ * This Dag Procedure will deal with pending Job statuses such as: PENDING, PENDING_RESUME, PENDING_RETRY
+ * as well jobs that have reached an end state with statuses such as: COMPLETED, FAILED and CANCELLED.
+ * Primarily, it will be responsible for polling the flow and job statuses and advancing to the next node in the dag.
+ *
+ */
+@Slf4j
+@Alpha
+public class AdvanceDagProc extends DagProc<AdvanceDagTask> {
+  private AdvanceDagTask advanceDagTask;
+  private Optional<Dag<JobExecutionPlan>> dagToAdvance;
+
+
+  public AdvanceDagProc(AdvanceDagTask advanceDagTask, DagProcFactory dagProcFactory) {
+    super(dagProcFactory);
+    this.advanceDagTask = advanceDagTask;
+  }
+
+  @Override
+  protected void initialize() throws IOException {
+    this.dagToAdvance = dagManagementStateStore.getDag(this.advanceDagTask.getAdvanceDagId().toString());
+  }
+
+  @Override
+  protected void act() {
+    if (!this.dagToAdvance.isPresent()) {
+      log.warn("No dag with id " + this.advanceDagTask.getAdvanceDagId() + " found to advance");
+      return;
+    }
+    // todo - find next dag node to run
+    Dag.DagNode<JobExecutionPlan> dagNode = new Dag.DagNode<>(new JobExecutionPlan(JobSpec.builder().build(), null));
```

Comment:
We should use dagId to retrieve DagNode from DagManagementStateStore (the next dagNode without a status)